### ISSUE TITLE
Very large performance regression from 0.22 to 0.24

### ISSUE THREAD

**Issue by jyn514** [servo/html5ever#414]
(I'm aware 0.25.1 is the latest version, if this is fixed on master let me know and I can close this. 0.25 removed RcDom so it would take some set up on my end to switch)

Docs.rs uses html5ever to parse HTML files generated by rustdoc. We recently updated from 0.22 to 0.24 and docs.rs started using so much CPU that all requests turned into 500 errors. The 0.24 version was deployed from 20:04 to 20:20 in the following diagram:

![image](https://user-images.githubusercontent.com/23638587/78072055-7718d600-736c-11ea-9ed9-16a2a894817f.png)

See also https://github.com/rust-lang/docs.rs/pull/669 (and many other PRs linked from there, we had to bisect in production).

Do you know what could have caused the slowdown? I saw that no new dependencies were introduced so it must have been something internal to html5ever. I can try to reproduce this on a single file if you're not sure, but right now it's preventing us from upgrading.

**Comment by jdm** [comment]
Can you share an example document that shows a measurable performance difference when parsing it?

**Comment by jdm** [comment]
My only guess as to what could have changed is https://github.com/servo/html5ever/commit/81ea581da593034b773de3c6f3e3535b0f66e51f; none of the rest of the commits look like they should have an impact on parsing performance.

**Comment by jyn514** [comment]
I can try, it might take me a while.

**Comment by jyn514** [comment]
It can't have been https://github.com/servo/html5ever/commit/81ea581da593034b773de3c6f3e3535b0f66e51f because that's only part of 0.25, not 0.24.

**Comment by jyn514** [comment]
Benchmarks with 0.25 and kuchiki still show a big slowdown, but not as bad as 0.24: https://github.com/rust-lang/docs.rs/pull/671#issuecomment-607514793

**Comment by jdm** [comment]
We probably won't be able to investigate this until sometime next week due to deadlines, unfortunately.

**Comment by jyn514** [comment]
Just checking on the status of this, you mentioned you might have a time to look at it this week? No problem if not, this isn't urgent for us.

**Comment by jdm** [comment]
We have not had any time due to the previous deadlines being extended to accommodate delays.

**Comment by jyn514** [comment]
Checking back in on this, let me know if there's any more I can do to identify the source of the regressions.

**Comment by jdm** [comment]
How am I supposed to run the benchmark code? The cratesfyi crate does not appear to exist on crates.io.

**Comment by jdm** [comment]
The most useful thing for me would be a standalone cargo project that runs a bunch of html5ever ever code infinitely in a loop that I can run in a profiler. Even better if it's very easy to swap out the version of html5ever that's in use in order to compare profiles and identify what changed to make it slower.

**Comment by jyn514** [comment]
Here is a standalone project that runs the html5ever code indefinitely in a loop: https://github.com/jyn514/html5ever-example

To swap out the html5ever, just change the version specified in `Cargo.toml`. Unfortunately this will only work for versions up to 0.24 (since rcdom was removed in 0.25). You can use 0.25 through kuchiki by checking out the `kuchiki` branch.

**Comment by jdm** [comment]
I applied the following patch to that example code:
```diff
diff --git a/src/main.rs b/src/main.rs
index 49c3787..95a25d1 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -8,9 +8,12 @@ type Result<T> = std::result::Result<T, failure::Error>;

 fn main() {
     let regex = include_str!("struct.CaptureMatches.html");
-    loop {
+    let now = std::time::Instant::now();
+    let iters = 50;
+    for _ in 1..iters {
         let (_head, _body, _classes) = extract_head_and_body(regex).unwrap();
     }
+    println!("{}ms/iter", (now.elapsed() / iters).as_millis());
 }
```
Using 0.24, I got `105ms/iter`. Using 0.23, I got `106ms/iter`. Using 0.22, I got `105ms/iter`. Before I spend time trying to dig in, I'd like to be sure that this code is actually representative of the performance regression that you're observing.

**Comment by jdm** [comment]
Thanks for the standalone project! I applied the following change to it:
```diff
diff --git a/src/main.rs b/src/main.rs
index 49c3787..4eda507 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -8,9 +8,12 @@ type Result<T> = std::result::Result<T, failure::Error>;

 fn main() {
     let regex = include_str!("struct.CaptureMatches.html");
-    loop {
+    let now = std::time::Instant::now();
+    let iters = 200;
+    for _ in 1..iters {
         let (_head, _body, _classes) = extract_head_and_body(regex).unwrap();
     }
+    println!("{}us/iter", (now.elapsed() / iters).as_micros());
 }

 /// Extracts the contents of the `<head>` and `<body>` tags from an HTML document, as well as the
```
And ran `cargo run --release` 8 times for each of 0.22, 0.23, and 0.24. I got these results:

```
0.22:
6799, 6952, 7042, 7056, 7078, 7096, 7135, 7272,

0.23:
7175, 7155, 7160, 7276, 7280, 7305, 7376, 7387, 

0.24:
7226, 7240, 7249, 7331, 7342, 7373, 7390, 7705, 
```
The spread definitely looks like versions since 0.22 take longer.

**Comment by jdm** [comment]
That being said, we're taking about a difference between 6.7ms and 7.7ms, but I'll see if a profile shows anything interesting.

**Comment by jyn514** [comment]
Awesome, thanks for the response! I agree that doesn't seem to explain the spike in usage we saw, but I'm at a loss as to what else it could be ... the only changes we made were those in https://github.com/rust-lang/docs.rs/pull/668/files (html5ever plus some transitive dependencies). I'll see whether we can try 0.25 instead and see if that has less of a usage spike than 0.24, I remember I benched it in https://github.com/rust-lang/docs.rs/pull/671#issuecomment-607514793 and it was slightly better.
